selfdrive/boardd/boardd.cc

```diff
@@ -644,48 +323,165 @@ void *can_recv_thread(void *crap) {
 
     next_frame_time += dt;
   }
-  return NULL;
 }
 
-void *can_health_thread(void *crap) {
+void can_health_thread() {
   LOGD("start health thread");
-  // health = 8011
   PubMaster pm({"health"});
 
+  uint32_t no_ignition_cnt = 0;
+  bool ignition_last = false;
+  float voltage_f = 12.5;  // filtered voltage
+
   // run at 2hz
   while (!do_exit) {
-    can_health(pm);
+    capnp::MallocMessageBuilder msg;
+    cereal::Event::Builder event = msg.initRoot<cereal::Event>();
+    event.setLogMonoTime(nanos_since_boot());
+    auto healthData = event.initHealth();
+
+    if (!panda || !panda->connected){
+      healthData.setHwType(cereal::HealthData::HwType::UNKNOWN);
+      pm.send("health", msg);
+      usleep(500*1000);
+      continue;
+    }
+
+    health_t health = panda->get_health();
+
+    if (spoofing_started) {
+      health.ignition_line = 1;
+    }
+
+    voltage_f = VOLTAGE_K * (health.voltage / 1000.0) + (1.0 - VOLTAGE_K) * voltage_f;  // LPF
+
+    // Make sure CAN buses are live: safety_setter_thread does not work if Panda CAN are silent and there is only one other CAN node
+    if (health.safety_model == (uint8_t)(cereal::CarParams::SafetyModel::SILENT)) {
+      panda->set_safety_model(cereal::CarParams::SafetyModel::NO_OUTPUT);
+    }
+
+    bool ignition = ((health.ignition_line != 0) || (health.ignition_can != 0));
+
+    if (ignition) {
+      no_ignition_cnt = 0;
+    } else {
+      no_ignition_cnt += 1;
+    }
+
+  #ifndef __x86_64__
```

Comment:
```suggestion
  #ifdef QCOM
```

This shouldn't apply to QCOM2 right?

Reply:
I think the power modes don't apply, but the safety model when ignition is off does apply. I'll add two conditions